approximate_string_matching/matching_with_dont_cares_and_mismatches.py

```diff
@@ -0,0 +1,284 @@
+import itertools
+import numpy as np
+from scipy.signal import convolve
+from scipy.stats import binom
+from approximate_string_matching import distance
+TOO_MANY_MISMATCHES = 'too many'
+
+def apply_preprocessing(func):
+  def wrapper(text, word, n, m, k, *args, **kwargs):
+    if m > n:
+      yield from ()
+    else:
+      # map characters to numbers, such that don't care is 0
+      alphabet = set(list(text[1:] + word[1:])) - set('?')
+      letter_mapping = {c: i for i, c in enumerate(alphabet, start = 1)}
+      letter_mapping.update({'?': 0})
+
+      # Since signal.convolve requires 0-indexing and we'll be working
+      # only on numbers from now, we'll remove sharps so
+      # text_numeric and word_numeric will be 0-indexed
+      text_numeric = np.array([letter_mapping.get(c) for c in text[1:]])
+      word_numeric = np.array([letter_mapping.get(c) for c in word[1:]])
+      yield from map(
+        lambda x: x+1,
+        func(text_numeric, word_numeric, n, m, k, *args, **kwargs)
+      )
+
+  return wrapper
+
+@apply_preprocessing
+def nonrecursive_randomised(text_numeric, word_numeric, n, m, k):
+  sample_rate = 1/max(k,1)
+
+  def random_masked_word_generator():
+    # O(k * log(n)), constant selected empirically with 2x safety threshold,
+    # which means, that c=15 also passes unit tests
+    for _ in range(int(k*(30+np.log(n)))):
+      yield np.array(
+        [x if np.random.random() < sample_rate else 0 for x in word_numeric]
+      )
+
+  return __nonrecursive_algorithm(
+    text_numeric, word_numeric, n, m, k, random_masked_word_generator)
+
+@apply_preprocessing
+def nonrecursive_deterministic(text_numeric, word_numeric, n, m, k):
+  ssf = generate_strongly_selective_family(m,k)
+
+  def ssf_masked_word_generator():
+    for s in ssf:
+      yield np.array(
+        [x if i in s else 0 for i,x in enumerate(word_numeric)]
+      )
+
+  return __nonrecursive_algorithm(
+    text_numeric, word_numeric, n, m, k, ssf_masked_word_generator)
+
+def __nonrecursive_algorithm(
+  text_numeric, word_numeric, n, m, k, masked_word_generator):
+
+  mismatches = [set() for _ in range(n-m+1)]
+  for masked_word in masked_word_generator():
+    found_mismatches = one_hamming_mismatch(text_numeric, masked_word)
+    for i, x in enumerate(found_mismatches):
+      if x is not None and x != TOO_MANY_MISMATCHES and len(mismatches[i]) < k:
+        mismatches[i].add(x)
+
+  yield from __checking_stage(text_numeric, word_numeric, mismatches)
+
+@apply_preprocessing
+def recursive(text_numeric, word_numeric, n, m, k):
+  mismatches = [set() for _ in range(n-m+1)]
+  E = [set() for _ in range(m)]
+
+  ks = k
+  while ks >= 1:
+    sample_rate = 1/ks
+    # O(ks + log(n)), constant selected empirically with 2x safety threshold,
+    # which means, that c=20 also passes unit tests
+    for _ in range(int(40*ks + np.log(n))):
```

Comment:
Good one but please extract the magic constant to the top of the function as KS_CONST (or with any other self-commenting name)

Reply:
Done, also the same for nonrecursive algorithm